apex/optimizers/fused_mixed_precision_lamb.py

```diff
@@ -0,0 +1,250 @@
+import torch
+from copy import deepcopy
+from itertools import chain
+from collections import defaultdict, abc as container_abcs
+
+from apex.multi_tensor_apply import multi_tensor_applier
+
+class FusedMixedPrecisionLamb(torch.optim.Optimizer):
+
+    def __init__(self, params, lr=1e-3, step=0, bias_correction=True,
+                 betas=(0.9, 0.999), eps=1e-6, weight_decay=0.01,
+                 amsgrad=False, adam_w_mode=True,
+                 grad_averaging=True, max_grad_norm=1.0, use_nvlamb=False,
+                 reduced_precision_dtype=None):
+        if amsgrad:
+            raise RuntimeError('FusedLAMB does not support the AMSGrad variant.')
+        
+        device = params[0].device
+
+        # The learning rate (lr) and optimizer step (step) should be located on device
+        # in order to faciliated device sync free execution
+        defaults = dict(lr=torch.tensor(lr, dtype=torch.float32, device=device),
+                        step=torch.tensor([step], dtype=torch.int, device=device),
+                        bias_correction=bias_correction,
+                        betas=betas, eps=eps, weight_decay=weight_decay,
+                        grad_averaging=grad_averaging,
+                        max_grad_norm=max_grad_norm)
+        super(FusedMixedPrecisionLamb, self).__init__(params, defaults)
+        if multi_tensor_applier.available:
+            import amp_C
+            self.multi_tensor_l2norm=amp_C.multi_tensor_l2norm_mp
+            # Skip buffer
+            self._dummy_overflow_buf = torch.tensor([0], dtype=torch.int, device=device)
+            self.multi_tensor_lamb = amp_C.multi_tensor_lamb_mp
+        else:
+            raise RuntimeError('apex.optimizers.FusedLAMB requires cuda extensions')
+
+        # Mixed Precision support
+        self.reduced_precision_dtype = reduced_precision_dtype
+        self.param_groups_full_precision = []
+        
+        self._step_supports_amp_scaling = True
+        self.adam_w_mode = 1 if adam_w_mode else 0
+        self.use_nvlamb = use_nvlamb
```

Comment:
So now that we have `param_groups`, we can move `lr` and `step` to GPU with an assumption that all param_groups are hosted on the same GPU device (the same assumption as the existing FusedLAMB I believe according to https://github.com/NVIDIA/apex/blob/aa756cec4359aff3df1d9abb68dc6e6e92920e0c/apex/optimizers/fused_lamb.py#L120).
```suggestion
        self.use_nvlamb = use_nvlamb
        
        # moving `lr` and `step` tensors to device as
        device = self.param_groups[0]["params"][0].device
        for group in self.param_groups:
            group["lr"] = group["lr"].to(device)
            group["step"] = group["step"].to(device)
```

Reply:
I just realized you put up this suggestion.  I added a similar one.